sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/WebExceptionRetryPolicy.java

```diff
@@ -41,79 +37,72 @@ public WebExceptionRetryPolicy(RetryContext retryContext) {
 
     @Override
     public Mono<ShouldRetryResult> shouldRetry(Exception e) {
-        if (isOutOfRetries) {
+        if (this.isOutOfRetries()) {
+            logger
+                .warn(
+                    "WebExceptionRetryPolicy() No more retrying on endpoint {}, operationType = {}, count = {}, " +
+                        "isAddressRefresh = {}",
+                    this.locationEndpoint,
+                    this.request.getOperationType(),
+                    this.retryCount,
+                    this.request.isAddressRefresh());
+
             this.durationTimer.stop();
             return Mono.just(ShouldRetryResult.noRetry());
         }
 
-        // Received Connection error (HttpRequestException), initiate the endpoint rediscovery
-        CosmosException webException = Utils.as(e, CosmosException.class);
-        if (WebExceptionUtility.isNetworkFailure(e) && this.isReadRequest &&
-            (webException != null && WebExceptionUtility.isReadTimeoutException(webException) &&
-                Exceptions.isSubStatusCode(webException, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_READ_TIMEOUT))) {
 
-            // if operationType AddressRefresh then just retry
-            if (this.request.isAddressRefresh()) {
-                return shouldRetryAddressRefresh();
+        if (WebExceptionUtility.isNetworkFailure(e)) {
+            if (this.isReadRequest || WebExceptionUtility.isWebExceptionRetriable(e)) {
```

Comment:
Actually we have an issue here, not introduced in this PR. The issue is:
AddressRefresh on httpTimeout is not being retried for write operations. Tracking item: https://github.com/Azure/azure-sdk-for-java/issues/37248